crates/turborepo-lib/src/globwatcher/mod.rs

```diff
@@ -0,0 +1,262 @@
+        // same output directories, however we are relying on task
+        // execution dependencies to prevent that.
+        self.config.flush().await.unwrap();
+
+        // hash_globs tracks all unchanged globs for a given hash.
+        // if a hash is not in globs, then either everything has changed
+        // or it was never registered. either way, we return all candidates
+        let hash_globs = self.hash_globs.lock().expect("only fails if poisoned");
+        match hash_globs.get(hash) {
+            Some(glob) => {
+                candidates.retain(|c| !glob.include.contains(c));
+                candidates
+            }
+            None => candidates,
+        }
+    }
+}
+
+/// iterate each path-glob pair and stop tracking globs whose files have
+/// changed. if a path is not a valid utf8 string, it is ignored. this is
+/// okay, because we don't register any paths that are not valid utf8,
+/// since the source globs are valid utf8
+///
+/// returns a list of hash-glob pairs to clear, and a list of globs to exclude
+///
+/// note: we take a mutex guard to make sure that the mutex is dropped
+///       when the function returns
+fn populate_hash_globs<'a>(
+    glob_statuses: &MutexGuard<HashMap<Glob, HashSet<Hash>>>,
+    repo_relative_paths: impl Iterator<Item = &'a Path> + Clone,
+    mut hash_globs: MutexGuard<HashMap<Hash, GlobSet>>,
+) -> (Vec<(Arc<String>, Arc<String>)>, Vec<Arc<String>>) {
+    let mut clear_glob_status = vec![];
+    let mut exclude_globs = vec![];
+
+    for ((glob, hash_status), path) in glob_statuses
+        .iter()
+        .cartesian_product(repo_relative_paths)
+        .filter(|((glob, _), path)| {
+            // ignore paths that don't match the glob, or are not valid utf8
+            path.to_str()
+                .map(|s| glob_match::glob_match(glob, s))
+                .unwrap_or(false)
+        })
+    {
+        // if we get here, we know that the glob has changed for every hash that
+        // included this glob and is not excluded by a hash's exclusion globs.
+        // So, we can delete this glob from every hash tracking it as well as stop
+        // watching this glob. To stop watching, we unref each of the
+        // directories corresponding to this glob.
+
+        for hash in hash_status.iter() {
+            let globs = match hash_globs.get_mut(hash).filter(|globs| {
+                globs
+                    .exclude
+                    .iter()
+                    .any(|f| glob_match::glob_match(f, path.to_str().unwrap()))
+            }) {
+                Some(globs) => globs,
+                None => continue,
+            };
+
+            // we can stop tracking that glob
+            globs.include.remove(glob);
+            if globs.include.is_empty() {
+                hash_globs.remove(hash);
+            }
+
+            // store the hash and glob so we can remove it from the glob_status
+            exclude_globs.push(glob.to_owned());
```

Comment:
Two comments on this line:
 1. I think `exclude_globs` will have duplicates, we could have multiple hashes per glob
 2. I _think_ we don't want to stop watching the glob if we hit the exclusion case above. 

Maybe we try something like `let mut stop_watching = true` before the loop over hashes, and we set `stop_watching = false` if we hit an exclusion. Then we can add to `exclude_globs` once if `stop_watching` is still true after the loop over hashes?